### PR TITLE
Feature/instantiate

### PR DESCRIPTION
**Pull request by maddyscientist**
* Introduced new `instantiate` helper to remove the boilerplate associated with instantiating the precision, color and reconstruct templates
* Any kernel that uses `instantiate` will automatically respect the `QUDA_PRECISION` and `QUDA_RECONSTRUCT` setting for optimized build times
* Applied this methodology to most of the "gauge kernels" with the notable exception of the gauge fixing and pure gauge kernels (these need extensive cleanup)
* This represents part of the continued evolution of making QUDA color agnostic
* Some minor fixes

### CODE REVIEW

**Comment by weinbe2** [comment]
@maddyscientist sorry this slipped, I'll look at this tomorrow. Do you know if the build failures are just because of other recent merges conflicting?

**Comment by maddyscientist** [comment]
Build failures are a known issue, and I'll push a fix for them (basically I have undefined behaviour in C++14, and the linker is failing on some builds and working for others).  Understood though, and fix will be easy.  I've been focussed on fixing the erroneous cast in release, hence I haven't gotten around to fixing this pull yet.

File: include/enum_quda_fortran.h [hunk 1]
```diff
@@ -45,14 +45,16 @@
...
-#define QUDA_BQCD_GAUGE_ORDER 10 //expect *gauge mu even-odd spacetime+halos row-column order
-#define QUDA_TIFR_GAUGE_ORDER 11
-#define QUDA_TIFR_PADDED_GAUGE_ORDER 12
+#define QUDA_NATIVE_GAUGE_ORDER 5 // used to denote one of the above types in a trait, not used directly
+#define QUDA_QDP_GAUGE_ORDER 6    // expect *gauge[4] even-odd spacetime row-column color
+#define QUDA_QDPJIT_GAUGE_ORDER 7 // expect *gauge[4] even-odd spacetime row-column color
+#define QUDA_CPS_WILSON_GAUGE_ORDER 8 // expect *gauge even-odd spacetime column-row color
+#define QUDA_MILC_GAUGE_ORDER 9       // expect *gauge even-odd mu spacetime row-column order
+#define QUDA_MILC_SITE_GAUGE_ORDER                                                                                     \
+  10                             // packed into MILC site AoS [even-odd][spacetime] array, and [dir][row][col] inside
```
Comment (weinbe2) [inline include/enum_quda_fortran.h]: This file must not get touched during a Jenkins test, because this certainly won't compile...

Comment (maddyscientist) [reply include/enum_quda_fortran.h]: Eventually, I would like to have this file auto generated from the c equivalent header to remove potential for errors like this.

File: include/instantiate.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      errorQuda("QUDA_PRECISION=%d does not enable double precision", QUDA_PRECISION);
+#endif
+    } else if (U.Precision() == QUDA_SINGLE_PRECISION) {
+#if QUDA_PRECISION & 4
+      instantiate<Apply, Recon, float>(U, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable single precision", QUDA_PRECISION);
+#endif
+    } else {
+      errorQuda("Unsupported precision %d\n", U.Precision());
```
Comment (weinbe2) [inline include/instantiate.h]: Why does this terminate after `single`? Alternatively, am I slightly missing where this is used, and the doxygen of `Additional arguments for different dslash kernels` isn't correct?

Comment (maddyscientist) [reply include/instantiate.h]: This helper is only used for gauge algorithms so we don’t need to have half or quarter precision support in there.

I will fix the doxgyen

File: lib/instantiate.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#include <instantiate.h>
+
+/**
+   This file contains deinitions required when compiling with C++14.
+   Without these, we can end up with undefined references at link
+   time.  We can remove this file when we jump to C++17.
+
+   For reasons I do not understand, this file fails to compile when
```
Comment (weinbe2) [inline lib/instantiate.cpp]: outdated comment

Comment (maddyscientist) [reply lib/instantiate.cpp]: Thx.  Will fix in next push

File: lib/llfat_quda.cu [hunk 4]
```diff
@@ -468,77 +406,43 @@ namespace quda {
...
-
-#endif //GPU_FATLINK
+  };
 
-  void fatLongKSLink(cudaGaugeField* fat, cudaGaugeField* lng,  const cudaGaugeField& u, const double *coeff)
+  // Compute the staple field for direction nu,excluding the directions dir1 and dir2.
+  void computeStaple(GaugeField &fat, GaugeField &staple, const GaugeField &mulink, const GaugeField &u,
+		     int nu, int dir1, int dir2, double coeff, bool save_staple)
   {
+    instantiate<Staple_, ReconstructNo12>(u, fat, staple, mulink, nu, dir1, dir2, coeff, save_staple);
```
Comment (weinbe2) [inline lib/llfat_quda.cu]: Add check to make sure link is `su3` and if not barf if recon 12 is requested

File: lib/llfat_quda.cu [hunk 5]
```diff
@@ -86,166 +93,97 @@ namespace quda {
...
-          errorQuda("Staggered phase type %d not supported", u.StaggeredPhase());
-        }
-      } else {
-	errorQuda("Reconstruct %d is not supported\n", u.Reconstruct());
-      }
-    } else {
-      errorQuda("Unsupported precision %d\n", u.Precision());
-    }
-    return;
+    instantiate<LongLink, ReconstructNo12>(u, lng, coeff); // u first arg so we pick its recon
```
Comment (weinbe2) [inline lib/llfat_quda.cu]: Add check that link is su(3) and if not barf if recon 12 is requested

File: lib/llfat_quda.cu [hunk 6]
```diff
@@ -86,166 +93,97 @@ namespace quda {
...
-      } else {
-        errorQuda("Reconstruct %d is not supported\n", u.Reconstruct());
-      }
-    } else {
-      errorQuda("Unsupported precision %d\n", u.Precision());
-    }
-    return;
+    if (u.StaggeredPhase() != QUDA_STAGGERED_PHASE_MILC && u.Reconstruct() != QUDA_RECONSTRUCT_NO)
+      errorQuda("Staggered phase type %d not supported", u.StaggeredPhase());
+    instantiate<OneLink, ReconstructNo12>(u, fat, coeff);
```
Comment (weinbe2) [inline lib/llfat_quda.cu]: Add su(3) check

File: include/instantiate.h (hunk 2 above)
Comment (weinbe2) [reply include/instantiate.h]: 👍

**Comment by weinbe2** [comment]
FYI to others, done with my review pending a few fixes from @maddyscientist that we discussed in the comments above

**Comment by maddyscientist** [comment]
I have pushed all the fix Evan noted on his review, and also fixed a slew of other fixes for bugs I found after extended testing today.  Some of these fixes will need to be cherry picked into release/1.0.x.

File: lib/gauge_field.cpp [hunk 7]
```diff
@@ -75,6 +76,11 @@ namespace quda {
       length = 2*2*nDim*stride*nInternal;  //two comes from being full lattice
     }
 
+    if ((reconstruct == QUDA_RECONSTRUCT_12 || reconstruct == QUDA_RECONSTRUCT_9) &&
```
Comment (weinbe2) [inline lib/gauge_field.cpp]: recon 9 -> recon 8

Comment (maddyscientist) [reply lib/gauge_field.cpp]: thank you 😄

Comment (weinbe2) [reply lib/gauge_field.cpp]: That error actually causes the staggered and improved staggered dslash tests to fail, so it's pretty important ;)

**Pull request merged**